accepted/2024/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,193 @@
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+A new NuGet CLI command option `stage` will be added to the `push` command, which will allow a user to stage a package to nuget, specifying a group identifier. (dotnet nuget push --stage [group-id])
+ Staged packages will be uploaded to NuGet and will be validated as part of the staging process, so that if a package being staged fails validation, the staging fails. After uploading multiple packages, the user can then call `nuget push --group [group-id]`
+ command to request all the packages from the earlier group to become available on the server as `published`.
+Note, that the `group` information will only be stored on the server and has nothing to do with how a package was built.
+
+Groups are scoped to an owner. So a package published by `Owner 1` into `Group 1` has nothing to do with another package published
+ to `Group 1` by `Owner 2`, as these should be treated as two separate groups.
+
+Groups can expand over time, by publishing new packages to the same group in the future. There is no concept of a "deprecated group"
+ as group by itself is stateless. It's just a collection of packages. A package can belong only to a single group.
+ This group expansion functionality will be handy for the error situation described above, where during staging a package staging fails.
+ This will result in only some packages being staged. The user will need to address the issues related to the failed package
+ and stage the remaining packages using the same command. Note, that if a package is staged, the operation should be idempotent,
+ and a request to stage the same package again should only result in a warning message (package is already staged) and the package shouldn't be uploaded again.
+
+Groups can later be used in the future when it's time to deprecate a set of packages. The package author can request to
+ deprecate a `group` by calling a single command on the nuget CLI and all the packages that belong to that group will be marked as `deprecated`.
+ This particular functionality depends on the support of the NuGet Deprecation command, which is not yet publicly available.
+
+With this information in place, new set of NuGet APIs / extensions will be needed to allow a user to run the following actions:
+- **Stage** a package for later publishing
+- **Publish an earlier staged group**
+- **Deprecate a group** using the group id 
+- **Discard a staged release** using the group id
```

Comment:
Should there be a capability to "unstage" a package? I.e. remove a package that haven't been published yet from a group?

Reply:
I haven't come across a scenario where that is needed.
If this will become a scenario that users need in the future, the feature set can be expanded to include that.
In the meantime, if customers will have to reset a stage, they will have to discard it and restage the updated set of packages.

Reply:
Scenario - someone accidentally staged a package. Or staged a package trying to figure out how that staging thing works.